Add `-e` to `run`, setting environment variables

lukaslueg opened the pull request:
Implement #4505 

A new option `-e` or `--env` is provided to `cargo run`. The process ultimately started has the given environment variables added to it. These environment variables are not visible to Cargo or rustc (think `LD_PRELOAD`). They are also not visible to `build.rs`, for which people may have a use-case. Personally, I'd rather wait until this comes up.

There may be a use-case for `cargo test`. Any thoughts?

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
Thanks for the PR! I've never personally been too huge of a fan of adding a feature like this, but all in all it's quite a small implementation and I think is consistent with other tools like `docker` (albeit maybe not quite as well motivated as with a tool like `docker`).

@rust-lang/cargo, any thoughts on this?

matklad commented:
We might want to have something like this for IntelliJ. Currently we don't execute binaries ourselves and use `cargo run`, so we have to set env vars for both Cargo and inferior process. But it actually works pretty well, we've never got a request to specify env vars only for executable itself. 

Also, using just `--env` would not allow to *remove* some environmental variables, which is required for the full generality. 

If we do add this flag, we should add it for `test` and `bench` as well I think.

All in all, I am neutral about this change: it's a niche usability improvement, which should be easy to support (although perhaps we can have only `--env` option? `-e` does not seem essential for a niche use-case, `env` is already short, and we might want to use `-e` for something else in the future).

lukaslueg commented:
I'll see to that

lukaslueg commented:
* Removed `-e` and left `--env`.
* The syntax `NAME=` will now clear the variable `NAME` in the target-process environment.
* Added `--env` to `cargo bench`. Only the benchmark binaries are affected by the newly set environment variables.
* Added `--env` to `cargo test`. This also includes doc-tests, which is a bit of a quirk, because `rustdoc` IS in fact affected by the newly set environment.

I refactored the argument parsing into a new function, for which I didn't find a nice spot to put in. There seems to be no common place for CLI-utility functions per in `cargo/util` or `/ops` so I just shared them between the bins. Any thoughts?

Please have a decent look if the tests do in fact cover all use-cases.

alexcrichton commented:
@matklad you bring up some good points! This is sort of precisely what I *don't* want to deal with, adding flags to remove env vars, ensuring we add it to all subcommands and plumb support, etc.

I'm not sure :(

withoutboats commented:
I lean against doing this. It's unfortunate that how envvars are set isn't consistent across platforms, but they do all allow you to set them. Without a clearer example of how bad the UX of setting env vars the usual way is (and maybe it just hasn't been explained well enough), I don't think its worth the maintenance and UX burden of providing additional flags to create a "cross platform" experience for setting env vars.

lukaslueg commented:
Consensus seems to be to not merge this. OP's original intention in #4505 was to make it possible to set environment variables for the target process without affecting Cargo or rustc itself during `cargo run`. Is #4505 to be closed essentially as WONTFIX or is there something to be salvaged ?

alexcrichton commented:
I'd personally consider #4505 as "wontfix" and instead pursue other means of fixing those related issues, yeah.

lukaslueg commented:
I'm fine with closing this then
